src/TemplateString.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+namespace ipl\Html;
+
+use Exception;
+
+/**
+ * Render {{#mustache}}Mustache{{/mustache}}-like string from {@link ValidHtml} element arguments
+ *
+ * # Example Usage
+ * ```
+ * $info = new TemplateString(
+ *     'Follow the %s for more information on %s {{#mustache}}Mustache{{/mustache}}',
+ *     [
+ *         new Link('doc/html', 'HTML documentation'),
+ *         Html::tag('strong', 'HTML elements'),
+ *        [
+ *        'mustache' => Html::tag('span')
+ *        ]
+ *     ]
+ * );
+ * ```
+ */
+class TemplateString extends FormattedString
+{
+    /** @var array */
+    protected $templateArgs = [];
+
+    protected $pos = 0;
+
+    protected $string;
+
+    protected $length;
+
+    protected $unclosed = [];
+
+    public function __construct($format, $args = null)
+    {
+        foreach ($args as $key => $val) {
+            if (is_array($val) && is_string(key($val))) {
+                unset($args[$key]);
+                $this->templateArgs += $val;
+            }
+        }
+
+        parent::__construct($format, $args);
+    }
+
+    /**
+     * @param null $for
+     * @return HtmlDocument
+     * @throws Exception in case of missing template argument or unbounded open or close templates
+     */
+    protected function parseTemplates($for = null)
+    {
+        $buffer = '';
+
+        while ($this->pos < $this->length) {
+            $char = $this->readChar();
```

Comment:
You could combine that like so:

```suggestion
        while (($char = $this->readChar()) !== false) {
```